Pull Request to Allow DNS on OpenVPN Reconnect

PacketShepard opened the pull request:
Hi Markus,

I decided to make a pull request to fix this issue. If OpenVPN experiences a connection failure and is using a FQDN as the remote server, the current container will not allow it to perform a DNS query to look up the IP address. This creates a permanent failure of OpenVPN unless you restart the container. It's caused by the leak prevention in iptables blocking DNS queries to ${NAME_SERVERS} on eth0 after the initial OpenVPN connection.

This could allow minor DNS leakage while OpenVPN reconnects, but I think is the best compromise for tunnel reliability. Please consider merging into master.

Thanks,

PacketShepard

PacketShepard committed:
Update README.md

Changed the readme to reflect the actual WEBUI_PORT variable used.

PacketShepard committed:
Update start.sh

Export name_server_list so it can be used in other scripts, such as the iptables script

PacketShepard committed:
Update iptables.sh - Added DNS Rules for OpenVPN

Added for loop logic to iterate through name_server_list array and add iptables rules for output DNS requests on eth0. This will allow OpenVPN to look up its remote server hostname on connection failure and reconnect. This DNS request should only work when the OpenVPN default route is lost. It is a compromise to have OpenVPN reliably be able to reconnect.

PacketShepard committed:
Update iptables.sh  - Move DNS Rules to One Stanza

Moved the iptables DNS rule calls to one stanza and fixed the incorrect interface syntax for output chain. Added /32 CIDR to the source and destination for completeness. Migrated over NAME_SERVERS readline split because array environment variable is not passing to the other shell.

claabs commented:
I'd appreciate this being merged. I currently have my container restart daily to avoid a long-term failure due to a short network outage.

PacketShepard committed:
Update start.sh - Updated default DNS

Updated default DNS to use CloudFlare and Quad9. Avoiding data mining and privacy issues with Google DNS as well as reliability issues with FreeDNS.

PacketShepard committed:
Update start.sh - Updated resolve.conf logic

Defaulting to override of the existing resolve conf. Docker-compose adds an internal resolver at the top of the file when an bridge driver network is created. It's run on loopback 127.0.0.11 and used to resolve the other containers in that docker-compose network. Non-local queries are forwarded to the host upstream resolver. OpenVPN doesn't handle the fail over to secondary and tertiary DNS well on ping reset so it will cause repeated failures if the nameserver 127.0.0.11 statement isn't removed.
See automatic DNS resolution here: https://docs.docker.com/network/bridge/

PacketShepard committed:
Update start.sh - Corrected Minor Typo with Resolv

Corrected "resolve" typo to "resolv" so that the resolv.conf is actually overwritten.

PacketShepard committed:
Update iptables.sh - Relax Iptables Int Req

Added comments above the iptables inserts for DNS lookups. Discovered in testing that OpenVPN keeps the default route while it's in a failure and reconnect state after a server times out. DNS then fails repeatedly even with the rules in place for eth0. Removed the interface requirement for eth0 in case the request is originating from the tun0 or loopback interfaces in this failure state. Stopping iptables seems to resolve the issue, so this rule change should fix the problem. Last resort is to relax the return DNS rule in case the IP of the DNS server isn't what's expected because of NAT or other docker weirdness.

PacketShepard committed:
Update Dockerfile

Added diagnostic utilities to the image, tcpdump and nslookup in dnsutils.
